src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
     end
     return dx
 end
+
+#####
+##### `foldl`
+#####
+
+# `foldl` guarantees to execute `f` in order, left to right. So it makes sense even when
+# this `f` is stateful, in which case the gradient must be calculated in the reverse order. 
+# The implementation aims to be efficient for both tuples and arrays. 
+# Note that it does not return a gradient for `init`.
+
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(foldl), f::F, x::Union{AbstractArray, Tuple};
+        init=_InitialValue()
+    ) where {F}
+    list, start = if init !== _InitialValue()
+        x, init
+    else
+        _drop1(x), first(x)
+    end
+    hobbits = accumulate(list; init=(start, pi)) do (a,_), b
+        c, back = rrule_via_ad(config, f, a, b)  # LHS is just documentation here
+        # Don't need to store every `c`, need last + one for the next iteration
+    end
+    y = first(last(hobbits))
+    axe = x isa AbstractArray ? axes(x) : nothing
+    type = Val(typeof(x))
+    project = x isa AbstractArray{<:Number} ? ProjectTo(x) : identity
+    function unfoldl(dy)
+        _tini = (pi, unthunk(dy), pi)
+        trio = accumulate(_reverse1(hobbits); init=_tini) do (_,dc,_), (_,back)
+            ds, da, db = back(dc)  # 's' for self, don't need to write LHS
+            # Don't need to store every `da`, need one for the next iteration + maybe last
+        end
+        df = sum(first, trio)
+        dx = map(t -> t[3], _reverse1(trio))
+        if init === _InitialValue()
+            # `hobbits` is one short
+            dx = _vcat1(trio[end][2], dx)
+        end
+        dx_nice = axe isa Tuple ? reshape(dx, axe) : Tangent{_val(type)}(dx...)
```

Comment:
This confused me a bit, til i realized this was another way of saying |

```suggestion
        dx_nice = _val(type) isa Tuple ?  Tangent{_val(type)}(dx...) : reshape(dx, axe)
```

How does this code look if we split the array case from the tuple case?

Can we teach `ProjectTo` to handle this for the tuple case?